beacon_node/store/src/hot_cold_store.rs

```diff
@@ -1580,157 +1744,226 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
+            storage_strategy
+                .replay_from_range(slot)
+                .rev()
+                .map(|prior_slot| historic_state_cache.get_state(prior_slot, &self.spec)),
+            |mut iter| iter.find_map(|cached_state| cached_state),
+        )?;
+        drop(historic_state_cache);
+
+        if let Some(cached_state) = cached_state {
+            if cached_state.slot() == slot {
+                metrics::inc_counter(&metrics::STORE_BEACON_HISTORIC_STATE_CACHE_HIT);
+                return Ok(cached_state);
             }
-            .map(Some)
-        } else {
-            Ok(None)
+            metrics::inc_counter(&metrics::STORE_BEACON_HISTORIC_STATE_CACHE_MISS);
+
+            return self.load_cold_state_by_slot_using_replay(cached_state, slot);
         }
-    }
 
-    /// Load a restore point state by its `state_root`.
-    fn load_restore_point(&self, state_root: &Hash256) -> Result<BeaconState<E>, Error> {
-        let partial_state_bytes = self
-            .cold_db
-            .get_bytes(DBColumn::BeaconState.into(), state_root.as_slice())?
-            .ok_or(HotColdDBError::MissingRestorePoint(*state_root))?;
-        let mut partial_state: PartialBeaconState<E> =
-            PartialBeaconState::from_ssz_bytes(&partial_state_bytes, &self.spec)?;
-
-        // Fill in the fields of the partial state.
-        partial_state.load_block_roots(&self.cold_db, &self.spec)?;
-        partial_state.load_state_roots(&self.cold_db, &self.spec)?;
-        partial_state.load_historical_roots(&self.cold_db, &self.spec)?;
-        partial_state.load_randao_mixes(&self.cold_db, &self.spec)?;
-        partial_state.load_historical_summaries(&self.cold_db, &self.spec)?;
-
-        let mut state: BeaconState<E> = partial_state.try_into()?;
-        state.apply_pending_mutations()?;
-        Ok(state)
+        metrics::inc_counter(&metrics::STORE_BEACON_HISTORIC_STATE_CACHE_MISS);
+
+        // Load using the diff hierarchy. For states that require replay we recurse into this
+        // function so that we can try to get their pre-state *as a state* rather than an hdiff
+        // buffer.
+        match self.hierarchy.storage_strategy(slot)? {
+            StorageStrategy::Snapshot | StorageStrategy::DiffFrom(_) => {
+                let buffer_timer =
+                    metrics::start_timer(&metrics::STORE_BEACON_HDIFF_BUFFER_LOAD_TIME);
+                let (_, buffer) = self.load_hdiff_buffer_for_slot(slot)?;
+                drop(buffer_timer);
+                let state = buffer.as_state(&self.spec)?;
+
+                self.historic_state_cache
+                    .lock()
+                    .put_both(slot, state.clone(), buffer);
+                Ok(state)
+            }
+            StorageStrategy::ReplayFrom(from) => {
+                let base_state = if let Some(state) = cached_state {
+                    // Found a prior state in the historic state cache.
+                    state
```

Comment:
`cached_state` can't be `Some` here, as it would have returned above.

Reply:
nice catch!

Reply:
fixed in https://github.com/sigp/lighthouse/pull/5978/commits/f6412f32b9df557dbe73d8aa430fc5b80da1c22a